### ISSUE TITLE
Proposal for c++ and matlab multiband implementation in the Gadgetron

### ISSUE THREAD

**Issue by valeryozenne** [gadgetron/gadgetron#636]
Hi everybody,

I would like to integrate the slice grappa reconstruction in the gadgetron framework.
I have a prototype version in Matlab that has been interfaced with the two regular sequences (the siemens product and the cmrr from minesotta version).
Before going futher, I would like to present a potential implementation.  
The information below do not present the actual slice grappa reco  (I will talk about in a other post) but just the steps that are necessary from my knowledge to reconstruct cartesian multiband data (either flash or epi) in a generic way.
Two gagdets "PrepGadget" & "PostGadget" deal with the different cases while the SliceGrappaGadget do only the reco.
Data could be transfered using the IsmrmrdReconData format with one additionnal dimension that correspond to the aliasing pattern/multiband pattern  [RO E1 E2 CHA N S SLC STACK] 
for instance 12 slices with slice acceleration of 3 become SLC=3 STACK=4 instead of SLC=12; 
This data will feed the SliceGrappaGadget that can be optimized to reconstruct all stacks in parallel using open-mp or gpu. 
Singleband data need to be identified from Multiband data, I suggest some minor modifications in the Gadgetron in order to differentiate reference data from parallel imaging (I mean grappa or sense) and from simultaneous multislice imaging (named single band or sb).

Here are below additionnal important modifications that maybe need to be validated before going futher.

### REQUIREMENTS:

the following modifications are required

- in ISMRMRD:
 
ismrmrd.h
add ISMRMRD_ACQ_SB (single band data) 
add ISMRMRD_ACQ_MB (multi band data)
add new xml and xls to grab the tag from siemens or cmrr sequence

- in Gadgetron:

in IsmrmrdReconBit
add boost::optional<IsmrmrdDataBuffered> sb_;

AcquisitionAccumulateTriggerGadget create a new one or change this one
add a new bucket for sb data that is called :" bucket->sbstats_ " // this is exactly the same code as for the data 

BucketToBufferGadget create a new one or change this one
fill IsmrmrdDataBuffered & dataBuffer = *rbit.sb_;  with the data // this is exactly the same code as for the data 

MatlabBufferGadget create a new one or change this one
const char * fieldname[3]= {"data", "reference", single"}
auto sb=mxGetField(bufferQ,idx,"single")
bit.sb_=MatlabStructBuffer(sb)

### GADGETS DESCRIPTIONS

------------------------------------------------

- GenericCartesianPrepMultibandGadget.cpp
- GenericCartesianPrepMultibandGadget.h

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% prepare data for slice-grappa reco

%% for all reformat data from 7D matrix to 8D matrix according to the multiband pattern in multiple stacks of MB slices

%% if EPI apply averaged ghost niquist correction
%% elseif flash do nothing
%% end

%% if CMRR sequence apply caipi 
%% else if siemens sequence apply specific phase shift 
%% end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Input kspace data using IsmrmrdReconData format  [RO E1 E2 CHA N S SLC]
Output kspace data using IsmrmrdReconData format  [RO E1 E2 CHA N S SLC STACK]

------------------------------------------------

- GenericCartesianSliceGrappaGadget.cpp
- GenericCartesianSliceGrappaGadget.h 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% do the slice grappa reco either Flash or EPI
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Class with heritage from GenericReconBase.cpp

use GenericReconCartesianSpiritObj which  record the recon kernel  for every encoding space

Input kspace data using IsmrmrdReconData format  [RO E1 E2 CHA N S SLC STACK] 
Output unliased kspace (in slice direction) data using IsmrmrdReconData format  [RO E1 E2 CHA N S SLC STACK] 

- GenericCartesianPostMultibandGadget.cpp
- GenericCartesianPostMultibandGadget.h

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% prepare data for futher grappa, partial fourier or simple FFT reco

%% if EPI undo apply average ghost niquist correction and apply optimal ghost niquist correction
%% elseif flash do nothing
%% end

%% if CMRR sequence unapply caipi shift in SB images 
%% else if siemens sequence 

%% for all reformat data from 8D matrix to 7D matrix according to the multiband pattern in multiple stacks of MB slices
%% for all undo blip pattern for MB images
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Best,

Valéry

**Comment by srapacchi** [comment]
Great idea Valéry. I do not see the need for a separate bucket for single band data, would the reference bucket not suffice for this purpose ? For SMS acquisitions, the reference should be single band, and can serve for reconstructing both the SMS undersampling and any in-plane undersampling.

Stan

**Comment by valeryozenne** [comment]
Hi Stan, yes I'm not 100% sure that we need to do it but I didn't manage to do it with an other way. 

For instance, using a 128x128 kspace with GRAPPA 2 and Slice acceleration of MB_factor ,  for both sequences in the first repetition I received:
- 24 * nb_slices lines ref  (ref)
- 64 lines * nb_slices  for  single band (sb)
- 64 lines * nb_slices/ MB_factor for multiband (mb)
then for each repetition 
- 64 lines * nb_slices/ MB_factor for multiband (mb)

In AcquisitionAccumulateTriggerMultibandGadget and BucketToBufferGadget, I need to differenciate single band from multiband data. If I do nothing the sb data will be mixed with the mb data as they share the same encoding pattern. One solution I found was to create a new bucket which allow to clearly separate ref, sb, and data (mb). 

If I put the data in the reference bucket, sb data will be also mixed with grappa reference lines (as 12  calibration lines share also the same encoding pattern than sb data) . It doesn't matter if the sample is fully static but in-vivo the lines from the grappa reference lines might differ a bit from the line from the sb data in some occasions.

Do you have any suggestions, I might miss a simple way to do it.

Valéry

**Comment by srapacchi** [comment]
Ok, this should be a private discussion maybe, but to my knowledge (i.e. manual from one SMS sequence), the single band data are optional, for quality control, and are not needed for the reconstruction. These data should follow the current pipeline, bypassing any SMS-specific gadget. Thus they should be part of the "data" buffer, not separate.

**Comment by dchansen** [comment]
Hi Valéry,

Sorry for not replying earlier. 
Having multiband recon in Gadgetron would certainly be great, but I'm a little concerned about the changes you propose.
Changing IsmrmrdReconBit is almost certainly not The Right Thing To Do (tm). I would propose using extra encoding spaces instead, or using a custom message type if need be. 
Do  you have a good description of how the MB recon works? That would help me understand how we could best engineer this. 

I am on holiday and paternity leave in the comming weeks, so I probably won't be able to do much before August, but I could read a few articles and put some thoughts into it.

**Comment by hansenms** [comment]
Stale, closing.
